File: CPP/Algospot/WILDCARD.cpp
#include <iostream>
#include <string>

std::string text, regex;
bool match = true;

void solve(int regex_index, int text_index) {
    char regex_char = regex.at(regex_index);
    switch (regex_char)
    {
        case '*': {
            while (regex_index < regex.length()) {
                switch (regex.at(regex_index))
                {
                    case '*': {
                        regex_index++;
                    } break;

                    case '\0':
                        return;
                    
                    case '?': {

                    } break;

                    default: {

                    } break;
                }
            }
        } break;

        case '?':
            return solve(regex_index + 1, text_index + 1);

        case '\0':
            return;

        default: {
            
        } break;
    }
}

int main() {
    int N;
    std::cin >> N;
    for(int _ = 0; _ < N; _++) {
        std::cin >> regex;

        int test_cases;
        std::cin >> test_cases;

        for(int i = 0; i < test_cases; i++) {
            std::cin >> text;
            solve(0, 0);
        }
    }

    return 0;
}

File: Python/algospot/WILDCARD.py
N = int(input())

def normalize(text):
    result = []
    wildcard = False
    for char in text:
        if char == '*':
            if not wildcard:
                wildcard = True

        else:
            if wildcard:
                result.append('*')
                wildcard = False
            result.append(char)

    if wildcard:
        result.append('*')

    return ''.join(result)

def solve(regex, text):
    index = 0
    while (index < len(regex) and index < len(text)) and (regex[index] == text[index] or regex[index] == '?'):
        index += 1

    if index == len(regex):
        return len(text) == index

    if regex[index] == '*':
        skip = 0
        while skip + index <= len(text):
            if (solve(regex[index + 1:], text[index + skip:])):
                return True

            skip += 1

        return False

    return False

for _ in range(N):
    regex = normalize(input())
    cases = int(input())

    results = []

    for i in range(cases):
        text = input()

        result = solve(regex, text)
        if (result):
            results.append(str(text))

    print('\n'.join(sorted(results)))

